### ISSUE TITLE
extend-protocol StreamSourcable for byte arrays

### ISSUE THREAD

**Issue by p-himik** [bbbates/clj-xml-validation#3]
Would be rather useful if `stream-source` could also be applied to byte arrays without having to wrap them.

**Comment by p-himik** [comment]
Immediate issues:
- Can't use `(class (byte-array 0))` in `extend-protocol` since it's not a symbol. It works if it's the very first type but it's probably not wise to rely on that
- Can't use `(Class/forName "[B")` for the same reason

But this answer contains an example of using a macro to overcome the issue: https://stackoverflow.com/a/27914334/564509

**Comment by p-himik** [comment]
Using
```clojure
(defmacro extend-byte-array-type [& specs]
  (let [ba (symbol "[B")]
    `(extend-type ~ba
       ~@specs)))

(extend-byte-array-type StreamSourcable
  (stream-source [b] (StreamSource. (io/input-stream b))))
```
appears to work just fine.

**Comment by bbbates** [comment]
`InputStream` implements `StreamSourcable` and, as you demo'd in your previous comment, it's super easy to create an input-stream from a byte array (calling `io/input-stream`). 
I'm not convinced that adding the snippet to extend the byte array type would make things any easier. Happy for you to try and convince me, though....

**Comment by p-himik** [comment]
The reasoning is exactly the same that I think could be applied to all other already implemented types: byte arrays are built-in, ubiquitous, and easy to implement. There's nothing different about them that calls for them to be left out.

**Comment by bbbates** [comment]
Not entirely true - the difference is that there isn't a `StreamSource` constructor that takes a ByteArray: https://docs.oracle.com/javase/8/docs/api/javax/xml/transform/stream/StreamSource.html
which is the reasoning behind the set of types that currently implement the `StreamSourcable` protocol.
I'm still not convinced, but if you think there is a strong need for it then raise a PR and we can look at it further.

**Comment by p-himik** [comment]
If that's the reasoning, why then the current implementation changes the semantics of the `String` constructor? Java uses `String` as a URI, clj-xml-validation uses `String` as a container for bytes.

**Comment by bbbates** [comment]
Yeah, good point. It was more around keeping the behaviour of the existing library that this was forked from.
Like I said: raise a PR and we can go from there.
